### DISCUSSION TITLE
What is the recommended way to refresh the `access_token`?

### DISCUSSION THREAD

**Discussion by elanzini** [supabase/supabase#9779] [Questions]
In my application, I use the `access_token` to identify the user when the client is making a request to the server, using the [`getUser()`](https://supabase.com/docs/reference/javascript/auth-getuser) call. 
I need the `access_token` on the client to be valid in order for this to work.

What is the recommended way to refresh access tokens on the client?

[This](https://stackoverflow.com/questions/72610091/supabase-rest-how-to-refresh-token) is the first result on Google but I was wondering if the `supabase-js` library already has something to handle this.

**Comment by GaryAustin1** [comment]
Supabase-js refreshes the jwt/access_token using a one time refresh token (also in the session) automatically by default on a browser after signIn.
This is a bit of info on the call to the Gotrue API used to exchange a refresh token for a jwt.  https://github.com/supabase/gotrue#post-token

**Reply by elanzini** [reply]
What I am seeing is a different behavior. If a user has already logged in, something is saved in the local storage and when I call `supabase.auth.getSession()`, I get the `access_token` present in this data. The first time the user logs in, I pass the jwt to the server and the server is able to get user information. After a while, the user opens the app and it looks like it's already logged in, but the token is not refreshed. So, when I try to `getUser()` info with the same jwt I get `null` data back.

**Reply by elanzini** [reply]
I have tried to use the endpoint you pointed to in order to refresh the token. To be clear, I am sending a `POST` request to my `supabase` instance at `/token?grant_type=referesh_token` passing the refresh token in the request body.

In `axios` it looks something like this:
```javascript
const body = {
  refresh_token: myRefreshToken,
};

axios
    .post(supabaseUrl + "/token?grant_type=refresh_token", body)
    .then(
      (response) => {
        console.log(response);
      },
      (error) => {
        console.log(error);
      }
    );
```

but I kept getting `401` error saying `No API key found in request` - `No apikey request header or url param was found.`
I tried to add the expired `access_token` as an Authorization Header but that didn't work either.

**Reply by GaryAustin1** [reply]
There is an apikey header that is normal your anon key.

**Reply by elanzini** [reply]
I inspected the calls that are made on the client while I use the `supabase-js` library - I'll try setting the `apiKey` header.

**Reply by elanzini** [reply]
I am now getting this issue: `Failed to load resource: Origin http://localhost:3000 is not allowed by Access-Control-Allow-Origin. Status code: 404`. I looked over the configs but could find anywhere to allow this origin.
